compiler/rustc_ast/src/util/classify.rs

```diff
@@ -51,8 +51,12 @@ pub fn expr_trailing_brace(mut expr: &ast::Expr) -> Option<&ast::Expr> {
             Gen(..) | Block(..) | ForLoop(..) | If(..) | Loop(..) | Match(..) | Struct(..)
             | TryBlock(..) | While(..) | ConstBlock(_) => break Some(expr),
 
-            // FIXME: These can end in `}`, but changing these would break stable code.
-            InlineAsm(_) | OffsetOf(_, _) | MacCall(_) | IncludedBytes(_) | FormatArgs(_) => {
+            MacCall(mac) => {
+                break (mac.args.delim == Delimiter::Brace).then_some(expr);
+            }
+
+            InlineAsm(_) | OffsetOf(_, _) | IncludedBytes(_) | FormatArgs(_) => {
```

Comment:
For the record, I don't even think we have access to the macro invocation's delimiters here, and it doesn't seem to be easy to thread them through from expansion. I'm pretty certain we don't care about this, though.

Reply:
:+1: I agree; this looks fine to me as you've implemented it.

Reply:
I am less sure of this - it should be possible to write the following:

```rust

let b"..." = include_bytes!(concat!(env!("OUT_DIR"), "/the_bytes")) else {
    return;
};
```

I suppose the difference here is that these are a compiler intrinsic macro?

I agree that it would be unlikely that some would write a statement like that with any of these though that is not better served by just an `if` expression.

(I am not sure what the rest of `break None;` is in the context for this match arm.)

Reply:
> it should be possible to write the following

It *is* possible to write that code. `break None;` means "this expression does not have a trailing brace".

Reply:
I provided an example that shows that this does not trigger on a parenthesized built-in macro: https://github.com/rust-lang/rust/pull/119062/files#diff-8a478faf29f76d7d6bff37c7d0f58aa2104a639d998936532774c7102c331152R167